File: backend/src/main/java/com/technicalescaperoom/backend/service/CinematicStoryService.java
package com.technicalescaperoom.backend.service;

import com.technicalescaperoom.backend.config.StorySequenceConfig;
import com.technicalescaperoom.backend.dto.story.ActiveStoryStateDto;
import com.technicalescaperoom.backend.dto.story.StorySequenceDto;
import com.technicalescaperoom.backend.entity.Player;
import com.technicalescaperoom.backend.entity.Team;
import com.technicalescaperoom.backend.entity.TeamStoryProgress;
import com.technicalescaperoom.backend.enums.StoryProgressStatus;
import com.technicalescaperoom.backend.exception.ResourceNotFoundException;
import com.technicalescaperoom.backend.repository.PlayerRepository;
import com.technicalescaperoom.backend.repository.TeamRepository;
import com.technicalescaperoom.backend.repository.TeamStoryProgressRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class CinematicStoryService {

    private final TeamRepository teamRepository;
    private final PlayerRepository playerRepository;
    private final TeamStoryProgressRepository teamStoryProgressRepository;
    private final StorySequenceConfig storySequenceConfig;
    private final GameWebSocketPublisher webSocketPublisher;

    @Transactional
    public ActiveStoryStateDto triggerStory(Long teamId, String storyKey) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new ResourceNotFoundException("Team not found: " + teamId));

        return triggerStory(team, storyKey);
    }

    @Transactional
    public ActiveStoryStateDto triggerStory(Team team, String storyKey) {
        if (team == null || storyKey == null) {
            return null;
        }

        // Idempotency: Has this story already been completed or skipped by this team?
        if (teamStoryProgressRepository.existsByTeamIdAndStoryKey(team.getId(), storyKey)) {
            log.debug("Story sequence {} already resolved for Team {}, ignoring duplicate trigger", storyKey, team.getTeamCode());
            return buildActiveState(team);
        }

        // Check if this exact story is already currently active
        if (storyKey.equals(team.getCurrentStoryKey())) {
            log.debug("Story sequence {} is already active for Team {}", storyKey, team.getTeamCode());
            return buildActiveState(team);
        }

        Optional<StorySequenceDto> seqOpt = storySequenceConfig.getSequence(storyKey);
        if (seqOpt.isEmpty()) {
            log.warn("Unknown story sequence key requested: {}", storyKey);
            return buildActiveState(team);
        }

        Instant now = Instant.now();
        team.setCurrentStoryKey(storyKey);
        team.setStoryPausedAt(now);
        teamRepository.save(team);

        TeamStoryProgress progress = TeamStoryProgress.builder()
                .team(team)
                .storyKey(storyKey)
                .status(StoryProgressStatus.ACTIVE)
                .startedAt(now)
                .pauseDurationSeconds(0L)
                .build();
        teamStoryProgressRepository.save(progress);

        ActiveStoryStateDto state = buildActiveState(team);
        log.info("▶ Story sequence [{}] triggered for Team {} (Timer PAUSED at {})", storyKey, team.getTeamCode(), now);
        webSocketPublisher.notifyStoryStarted(team.getId(), storyKey, state);

        return state;
    }

    @Transactional
    public ActiveStoryStateDto skipStory(Long teamId, Long playerId) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new ResourceNotFoundException("Team not found: " + teamId));

        if (!team.isStoryActive()) {
            log.debug("No active story to skip for Team {}", team.getTeamCode());
            return buildActiveState(team);
        }

        String storyKey = team.getCurrentStoryKey();
        Instant pausedAt = team.getStoryPausedAt();
        Instant now = Instant.now();

        long pauseDuration = (pausedAt != null) ? Math.max(0, Duration.between(pausedAt, now).getSeconds()) : 0L;
        team.setTotalStoryPauseSeconds(team.getTotalStoryPauseSeconds() + pauseDuration);
        team.setCurrentStoryKey(null);
        team.setStoryPausedAt(null);
        teamRepository.save(team);

        Optional<TeamStoryProgress> progOpt = teamStoryProgressRepository.findByTeamIdAndStoryKey(teamId, storyKey);
        if (progOpt.isPresent()) {
            TeamStoryProgress prog = progOpt.get();
            prog.setStatus(StoryProgressStatus.SKIPPED);
            prog.setEndedAt(now);
            prog.setPauseDurationSeconds(pauseDuration);
            teamStoryProgressRepository.save(prog);
        }

        Integer playerNum = null;
        if (playerId != null) {
            playerNum = playerRepository.findById(playerId).map(Player::getPlayerNumber).orElse(null);
        }

        ActiveStoryStateDto state = buildActiveState(team);
        log.info("⏩ Story sequence [{}] SKIPPED for Team {} by Operator {} (Paused for {}s, Timer RESUMED)",
                storyKey, team.getTeamCode(), (playerNum != null ? playerNum : "Unknown"), pauseDuration);

        webSocketPublisher.notifyStorySkipped(teamId, playerId, playerNum, storyKey, state);
        return state;
    }

    @Transactional
    public ActiveStoryStateDto completeStory(Long teamId, Long playerId) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new ResourceNotFoundException("Team not found: " + teamId));

        if (!team.isStoryActive()) {
            log.debug("No active story to complete for Team {}", team.getTeamCode());
            return buildActiveState(team);
        }

        String storyKey = team.getCurrentStoryKey();
        Instant pausedAt = team.getStoryPausedAt();
        Instant now = Instant.now();

        long pauseDuration = (pausedAt != null) ? Math.max(0, Duration.between(pausedAt, now).getSeconds()) : 0L;
        team.setTotalStoryPauseSeconds(team.getTotalStoryPauseSeconds() + pauseDuration);
        team.setCurrentStoryKey(null);
        team.setStoryPausedAt(null);
        teamRepository.save(team);

        Optional<TeamStoryProgress> progOpt = teamStoryProgressRepository.findByTeamIdAndStoryKey(teamId, storyKey);
        if (progOpt.isPresent()) {
            TeamStoryProgress prog = progOpt.get();
            prog.setStatus(StoryProgressStatus.COMPLETED);
            prog.setEndedAt(now);
            prog.setPauseDurationSeconds(pauseDuration);
            teamStoryProgressRepository.save(prog);
        }

        Integer playerNum = null;
        if (playerId != null) {
            playerNum = playerRepository.findById(playerId).map(Player::getPlayerNumber).orElse(null);
        }

        ActiveStoryStateDto state = buildActiveState(team);
        log.info("✔ Story sequence [{}] COMPLETED for Team {} (Paused for {}s, Timer RESUMED)",
                storyKey, team.getTeamCode(), pauseDuration);

        webSocketPublisher.notifyStoryCompleted(teamId, playerId, playerNum, storyKey, state);

        // Chain Prologue directly into Level 1 Intro narrative if not yet resolved
        if ("STORY_PROLOGUE".equals(storyKey) && !teamStoryProgressRepository.existsByTeamIdAndStoryKey(teamId, "STORY_L1_INTRO")) {
            return triggerStory(team, "STORY_L1_INTRO");
        }

        return state;
    }

    @Transactional
    public ActiveStoryStateDto replayStory(Long teamId, String storyKey) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new ResourceNotFoundException("Team not found: " + teamId));

        if (storyKey == null) {
            return buildActiveState(team);
        }

        Optional<StorySequenceDto> seqOpt = storySequenceConfig.getSequence(storyKey);
        if (seqOpt.isEmpty()) {
            log.warn("Unknown story sequence key for replay: {}", storyKey);
            return buildActiveState(team);
        }

        Instant now = Instant.now();
        team.setCurrentStoryKey(storyKey);
        team.setStoryPausedAt(now);
        teamRepository.save(team);

        ActiveStoryStateDto state = buildActiveState(team);
        log.info("▶ Replaying story sequence [{}] for Team {}", storyKey, team.getTeamCode());
        webSocketPublisher.notifyStoryStarted(team.getId(), storyKey, state);
        return state;
    }

    @Transactional(readOnly = true)
    public ActiveStoryStateDto getCurrentStoryState(Long teamId) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new ResourceNotFoundException("Team not found: " + teamId));

        return buildActiveState(team);
    }

    @Transactional(readOnly = true)
    public List<String> getResolvedStoryKeys(Long teamId) {
        return teamStoryProgressRepository.findByTeamIdOrderByStartedAtAsc(teamId).stream()
                .map(TeamStoryProgress::getStoryKey)
                .collect(Collectors.toList());
    }

    public long calculateEffectiveActiveDuration(Team team) {
        Instant startTime = (team.getStartedAt() != null) ? team.getStartedAt() : team.getCreatedAt();
        Instant endTime = (team.getCompletedAt() != null) ? team.getCompletedAt() : Instant.now();
        if (startTime == null) return 0L;

        long grossSeconds = Math.max(0, Duration.between(startTime, endTime).getSeconds());
        long totalPauses = (team.getTotalStoryPauseSeconds() != null) ? team.getTotalStoryPauseSeconds() : 0L;

        long inFlightPause = 0L;
        if (team.isStoryActive() && team.getStoryPausedAt() != null && team.getCompletedAt() == null) {
            inFlightPause = Math.max(0, Duration.between(team.getStoryPausedAt(), Instant.now()).getSeconds());
        }

        return Math.max(0, grossSeconds - totalPauses - inFlightPause);
    }

    private ActiveStoryStateDto buildActiveState(Team team) {
        boolean active = team.isStoryActive();
        String currentKey = team.getCurrentStoryKey();
        Instant pausedAt = team.getStoryPausedAt();

        Long currentPause = null;
        if (active && pausedAt != null) {
            currentPause = Math.max(0, Duration.between(pausedAt, Instant.now()).getSeconds());
        }

        StorySequenceDto seq = null;
        if (active && currentKey != null) {
            seq = storySequenceConfig.getSequence(currentKey).orElse(null);
        }

        List<String> resolvedKeys = teamStoryProgressRepository.findByTeamIdOrderByStartedAtAsc(team.getId()).stream()
                .filter(p -> p.getStatus() == StoryProgressStatus.COMPLETED || p.getStatus() == StoryProgressStatus.SKIPPED)
                .map(TeamStoryProgress::getStoryKey)
                .collect(Collectors.toList());

        return ActiveStoryStateDto.builder()
                .isStoryActive(active)
                .storyKey(currentKey)
                .storyPausedAt(pausedAt)
                .currentPauseSeconds(currentPause)
                .totalStoryPauseSeconds(team.getTotalStoryPauseSeconds() != null ? team.getTotalStoryPauseSeconds() : 0L)
                .effectiveActiveDurationSeconds(calculateEffectiveActiveDuration(team))
                .sequence(seq)
                .completedStoryKeys(resolvedKeys)
                .build();
    }
}

File: backend/services/cinematicStoryService.js
const teamRepository = require('../repositories/teamRepository');
const teamStoryProgressRepository = require('../repositories/teamStoryProgressRepository');
const storyConfig = require('../config/storyConfig');
const webSocketService = require('./webSocketService');
const { ResourceNotFoundException } = require('../middleware/errorHandler');

class CinematicStoryService {
  buildActiveState(team) {
    const hasActiveStory = Boolean(team.currentStoryKey);
    const sequence = hasActiveStory ? storyConfig.getSequence(team.currentStoryKey) : null;

    return {
      hasActiveStory,
      storyKey: team.currentStoryKey || null,
      title: sequence ? sequence.title : null,
      subTitle: sequence ? sequence.subTitle : null,
      canSkip: sequence ? sequence.canSkip : true,
      lines: sequence ? sequence.lines : [],
      pausedAt: team.storyPausedAt || null,
      totalPauseSeconds: team.totalStoryPauseSeconds || 0
    };
  }

  async triggerStory(teamOrId, storyKey) {
    let team = typeof teamOrId === 'object' ? teamOrId : await teamRepository.findById(teamOrId);
    if (!team || !storyKey) return null;

    // Idempotency check: has this story already been completed or skipped?
    const alreadyResolved = await teamStoryProgressRepository.existsByTeamIdAndStoryKey(team.id, storyKey);
    if (alreadyResolved) {
      return this.buildActiveState(team);
    }

    if (storyKey === team.currentStoryKey) {
      return this.buildActiveState(team);
    }

    const seq = storyConfig.getSequence(storyKey);
    if (!seq) {
      return this.buildActiveState(team);
    }

    const now = new Date().toISOString();
    team.currentStoryKey = storyKey;
    team.storyPausedAt = now;
    await teamRepository.save(team);

    await teamStoryProgressRepository.save({
      teamId: team.id,
      storyKey,
      status: 'ACTIVE',
      pauseDurationSeconds: 0
    });

    const state = this.buildActiveState(team);
    webSocketService.notifyStoryStarted(team.id, storyKey, state);
    return state;
  }

  async skipStory(teamId, playerId) {
    const team = await teamRepository.findById(teamId);
    if (!team) throw new ResourceNotFoundException('Team not found');

    if (!team.currentStoryKey) {
      return this.buildActiveState(team);
    }

    const storyKey = team.currentStoryKey;
    let pauseSec = 0;
    if (team.storyPausedAt) {
      pauseSec = Math.max(0, Math.floor((Date.now() - new Date(team.storyPausedAt).getTime()) / 1000));
      team.totalStoryPauseSeconds = (team.totalStoryPauseSeconds || 0) + pauseSec;
    }

    team.currentStoryKey = null;
    team.storyPausedAt = null;
    await teamRepository.save(team);

    const progress = await teamStoryProgressRepository.findByTeamIdAndStoryKey(team.id, storyKey);
    if (progress) {
      progress.status = 'SKIPPED';
      progress.completedAt = new Date().toISOString();
      progress.completedByPlayerId = playerId;
      progress.pauseDurationSeconds = pauseSec;
      await teamStoryProgressRepository.save(progress);
    }

    webSocketService.notifyStoryCompleted(team.id, storyKey);
    return this.buildActiveState(team);
  }

  async completeStory(teamId, playerId) {
    const team = await teamRepository.findById(teamId);
    if (!team) throw new ResourceNotFoundException('Team not found');

    if (!team.currentStoryKey) {
      return this.buildActiveState(team);
    }

    const storyKey = team.currentStoryKey;
    let pauseSec = 0;
    if (team.storyPausedAt) {
      pauseSec = Math.max(0, Math.floor((Date.now() - new Date(team.storyPausedAt).getTime()) / 1000));
      team.totalStoryPauseSeconds = (team.totalStoryPauseSeconds || 0) + pauseSec;
    }

    team.currentStoryKey = null;
    team.storyPausedAt = null;
    await teamRepository.save(team);

    const progress = await teamStoryProgressRepository.findByTeamIdAndStoryKey(team.id, storyKey);
    if (progress) {
      progress.status = 'COMPLETED';
      progress.completedAt = new Date().toISOString();
      progress.completedByPlayerId = playerId;
      progress.pauseDurationSeconds = pauseSec;
      await teamStoryProgressRepository.save(progress);
    }

    webSocketService.notifyStoryCompleted(team.id, storyKey);
    return this.buildActiveState(team);
  }

  async replayStory(teamId, storyKey) {
    const team = await teamRepository.findById(teamId);
    if (!team) throw new ResourceNotFoundException('Team not found');

    const seq = storyConfig.getSequence(storyKey);
    if (!seq) {
      return this.buildActiveState(team);
    }

    // Passive viewing: do not pause game timer for replays
    team.currentStoryKey = storyKey;
    team.storyPausedAt = null;
    await teamRepository.save(team);

    return this.buildActiveState(team);
  }

  async getResolvedStoryKeys(teamId) {
    return teamStoryProgressRepository.findResolvedByTeamId(teamId);
  }

  async getCurrentStoryState(teamId) {
    const team = await teamRepository.findById(teamId);
    if (!team) throw new ResourceNotFoundException('Team not found');
    return this.buildActiveState(team);
  }
}

module.exports = new CinematicStoryService();
